recipe/patch_yaml/cupy.yaml

```
@@ -15,3 +15,12 @@ then:
   - replace_constrains:
       old: libcugraph >=0.19.0,<1.0a0
       new: libcugraph >=0.19.0
+---
+if:
+  name: cupy-core
+  version: 13.0.0
+  build_number_in: [0, 1, 2, 3]
+  timestamp_lt: 1708656816000
+then:
+  # create conflicts with older cupy to avoid coexistence
+  - add_constrains: cupy<13.0a0
```

```suggestion
  - add_constrains: cupy=${version}
```
We want to add the following to all the previously published `cupy-core` packages.
```yaml
requirements:
  run_constrained:
    - cupy ={{ version }}
```
because `cupy-core` should only be installed with a matching `cupy` version?

---

1. `cupy-core` exists starting v13.0.0. The current patch covers all of them. 
2. The constraint that you suggested (only installed with a matching `cupy`) was already enforced at [this line](https://github.com/conda-forge/cupy-feedstock/blob/c08b8c624246a95a9381e5c8a0483cbe34d446a5/recipe/meta.yaml#L52), since the beginning of `cupy-core`. So I want this patch to be very explicit about why it failed, and once we exclude the failure factors, the current recipe kicks in and does the rest of the job right.

---

1. Yes.
2. No. This patch is for the metadata of `cupy-core`. The line you referenced is a runtime constraint for `cupy` (not `cupy-core`). It's good that `cupy` depends on a specific version of `cupy-core`, but `cupy-core` also needs to depend on a specific version of `cupy`. That's the purpose of this PR, to make `cupy-core` depend on a specific version of `cupy` so it cannot be installed with old versions of cupy.

---

12 < 13.0a0 < 13.0.0 ? Right?

If `cupy-core` depends on `cupy <13.0a0`, then `cupy =13.*` can never be installed because it requires a `cupy-core` of the same version.

https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#predictable-version-ordering

---

Think we want more changes than this change then. Have proposed a few

---

I think that:

```
requirements:
  run_constrained:
    - cupy ={{ version }}
```

should be safe enough...
no need to complicate things too much